### PR TITLE
Implement recursive and memory test in go-owasm

### LINKED ISSUE #1990: Implement recursive test case for go-wasm

### LINKED ISSUE #1991: Implement test about memory for go-owasm

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #1990, #1991

### CODE REVIEW

File: go-owasm/api/lib_test.go [hunk 1]
```diff
@@ -278,3 +278,108 @@ func TestCompileErrorCheckWasmExports(t *testing.T) {
...
+		(memory 17)
+		(export "prepare" (func 1))
+		(export "execute" (func 2)))
+
+		`), spanSize)
+	code, err := Compile(wasm, spanSize)
+	require.NoError(t, err)
+
+	err = Prepare(code, 100000, NewMockEnv([]byte("")))
+	require.Equal(t, ErrUnknownError, err)
```
Comment (sorawit) [inline go-owasm/api/lib_test.go]: We should not knowingly have unknown error.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: go-owasm/api/lib_test.go [hunk 2]
```diff
@@ -278,3 +278,108 @@ func TestCompileErrorCheckWasmExports(t *testing.T) {
...
+		(memory 17)
+		(export "prepare" (func 1))
+		(export "execute" (func 2)))
+
+		`), spanSize)
+	code, err := Compile(wasm, spanSize)
+	require.NoError(t, err)
+
+	err = Prepare(code, 100000, NewMockEnv([]byte("")))
+	require.Equal(t, ErrCompliationFail, err)
```
Comment (sorawit) [inline go-owasm/api/lib_test.go]: Let's change to instantiate error.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

File: go-owasm/api/error.go [hunk 3]
```diff
@@ -72,6 +73,8 @@ func toGoError(code C.Error) error {
 		return ErrGasCounterInjection
 	case C.Error_StackHeightInjectionError:
 		return ErrStackHeightInjection
+	case C.Error_OutOfMemoryRangeError:
```
Comment (sorawit) [inline go-owasm/api/error.go]: Let's move this to the right place

File: go-owasm/api/bindings.h [hunk 4]
```diff
@@ -44,14 +54,14 @@ typedef struct env_t {
...
 typedef struct EnvDispatcher {
-  Error (*get_calldata)(env_t*, Span *calldata);
-  Error (*set_return_data)(env_t*, Span data);
-  int64_t (*get_ask_count)(env_t*);
-  int64_t (*get_min_count)(env_t*);
-  Error (*get_ans_count)(env_t*, int64_t*);
-  Error (*ask_external_data)(env_t*, int64_t eid, int64_t did, Span data);
-  Error (*get_external_data_status)(env_t*, int64_t eid, int64_t vid, int64_t *status);
-  Error (*get_external_data)(env_t*, int64_t eid, int64_t vid, Span *data);
+  Error (*get_calldata)(env_t *, Span *calldata);
```
Comment (sorawit) [inline go-owasm/api/bindings.h]: Make sure to use cbindgen format

File: go-owasm/api/bindings.h [hunk 5]
```diff
@@ -24,6 +24,7 @@ enum Error {
   Error_RuntimeError = 11,
   Error_OutOfGasError = 12,
   Error_BadEntrySignatureError = 13,
+  Error_OutOfMemoryRangeError = 14,
```
Comment (sorawit) [inline go-owasm/api/bindings.h]: MemoryOutOfBoundError

File: go-owasm/api/error.go [hunk 6]
```diff
@@ -20,6 +20,7 @@ var (
 	ErrRuntime                 = errors.New("runtime error while executing the Wasm script")
 	ErrOutOfGas                = errors.New("out-of-gas while executing the wasm script")
 	ErrBadEntrySignature       = errors.New("bad execution entry point sigature")
+	ErrOutOfMemoryRange        = errors.New("out of memory range")
```
Comment (sorawit) [inline go-owasm/api/error.go]: out-of-bound memory access while executing the wasm script

File: go-owasm/src/error.rs [hunk 7]
```diff
@@ -18,6 +18,7 @@ pub enum Error {
   RuntimeError = 11,           // Runtime error while executing the Wasm script.
   OutOfGasError = 12,          // Out-of-gas while executing the Wasm script.
   BadEntrySignatureError = 13, // Bad execution entry point sigature.
+  OutOfMemoryRangeError = 14,  // Request out of memory range.
```
Comment (sorawit) [inline go-owasm/src/error.rs]: Update this comment accordingly.

File: go-owasm/src/lib.rs [hunk 8]
```diff
@@ -161,21 +168,22 @@ fn run(code: &[u8], gas_limit: u32, span_size: i64, is_prepare: bool, env: Env)
...
                 let vm: &mut vm::VMLogic = unsafe { &mut *(ctx.data as *mut vm::VMLogic) };
-                let span_size = vm.get_span_size() as usize;
-                // TODO: span_size bound check
-                let mut mem: Vec<u8> = Vec::with_capacity(span_size);
-                let mut calldata = Span::create_writable(mem.as_mut_ptr(), span_size);
+                let span_size = vm.get_span_size();
+                let mut mem: Vec<u8> = Vec::with_capacity(span_size as usize);
+                let mut calldata = Span::create_writable(mem.as_mut_ptr(), span_size as usize);
                 vm.get_calldata(&mut calldata)?;
+                require_mem_range( ctx.memory(0).size().bytes().0, (ptr +span_size) as usize )?;
```
Comment (sorawit) [inline go-owasm/src/lib.rs]: Let's move this check to right after when `span_size` is created. also make sure code is properly formatted.

ditto for all places below.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Comment by taobun** [comment]
Add change log

**Pull request merged**